src/Parser/Parser.hs

```diff
@@ -236,16 +236,25 @@ atom' =
   <|>
   Tuple <$> parens (commaSep1 expr)
   <|>
-  Let <$> (reserved "let" *> identifier) <*> (reservedOp "=" *> expr) <*> (reserved "in" *> expr)
+  (do
+      reserved "let"
+      var <- identifier
+      reservedOp "="
+      outer <- expr
+      reserved "in"
+      putWhileNames [var]
+      inner <- expr
+      return $ Let var outer inner
+  )
   <|>
   If <$> (reserved "if" *> expr) <*> (reserved "then" *> expr) <*> (reserved "else" *> expr)
   <|>
   (do
       reserved "while"
+      names <- getWhileNames
```

Comment:
The condition may contain a let expression. However, that should not alter the local state of this while loop. Getting the names here prevents that from occurring.